cfme/utils/ssh.py

```diff
@@ -842,3 +852,50 @@ def keygen():
     with _ssh_pubkey_file.open('w') as f:
         f.write("{} {} {}\n".format(pub.get_name(), pub.get_base64(),
             'autogenerated cfme_tests key'))
+
+
+@contextmanager
+def connect_ssh(vm, creds,
+                num_sec=CONNECT_RETRIES_TIMEOUT,
+                connect_timeout=CONNECT_TIMEOUT,
+                delay=CONNECT_SSH_DELAY):
+    """
+    A context manager attemptinng to connect all the IPs of the vm in several
+    rounds. After each round it will delay.
+    Once connective IP is found, it will return SSHClient with established
+    connection.
+
+    When returned value gones out of scope, the connection will be closed.
+    """
+
+    msg = "The num_sec is smaller then connect_timeout. This looks like a bug."
+    assert num_sec > connect_timeout, msg
+
+    timestamp = time.monotonic()
+    # Note we do need to be sure we get the fresh list of IPs for every cycle.
+    # The provider may be updating it.
+    for ip in _trying_fresh_ips(vm, delay=delay, rounds=-1):
+        # Timeout handling.
+        total_duration = (time.monotonic() - timestamp)
+        if total_duration > num_sec:
+            raise TimedOutError(f"None of the ips of vm {vm} were found "
+                f"connectible after trying each in {total_duration:.0f} seconds "
+                "total.")
+        try:
+            with SSHClient(hostname=ip, username=creds.principal, password=creds.secret,
+                           use_check_port=False, timeout=connect_timeout) as client:
+                logger.info(f"SSH connected to IP {ip}")
+
+                # This command execution check is probably redundant.
```

Comment:
I agree, I thought this was redundant the first time I saw it but didn't know if there was some other reason for it that I was unaware.